src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             }
 
+            if (!TryWriteValueToLocation(value, hashTableLocal, tableIndex))
+            {
+                // Indicates that the overwrite of the sentinel didn't work. That shouldn't be possible, so failfast
+                Environment.FailFast("Unable to overwrite sentinel with value");
+            }
+
             // If the write succeeded, increment _count
             Interlocked.Increment(ref _count);
             return value;
         }
 
         /// <summary>
-        /// Attampts to write a value into the table. May fail if another value has been added.
+        /// Attempts to write a the sentinel into the table. May fail if another value has been added.
+        /// </summary>
+        /// <returns>True if the sentinel was successfully written</returns>
+        private bool TryWriteSentinelToLocation(TValue[] hashTableLocal, int tableIndex)
+        {
+            // Add to hash, use a volatile write to ensure that
+            // the contents of the value are fully published to all
+            // threads before adding to the hashtable
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], _entryInProcessOfWritingSentinel, null) == null)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to write a value into the table. Should never fail as the sentinel should be the only
+        /// entry that can be in the table at this point
         /// </summary>
         /// <returns>True if the value was successfully written</returns>
         private bool TryWriteValueToLocation(TValue value, TValue[] hashTableLocal, int tableIndex)
         {
             // Add to hash, use a volatile write to ensure that
             // the contents of the value are fully published to all
             // threads before adding to the hashtable
-            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, null) == null)
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], value, _entryInProcessOfWritingSentinel) == _entryInProcessOfWritingSentinel)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to abort write a value into the table. Should never fail as the sentinel should be the only
+        /// entry that can be in the table at this point
+        /// </summary>
+        /// <returns>True if the value was successfully written</returns>
+        private bool TryWriteAbortNullToLocation(TValue[] hashTableLocal, int tableIndex)
+        {
+            // Add to hash, use a volatile write to ensure that
+            // the contents of the value are fully published to all
+            // threads before adding to the hashtable
+            if (Interlocked.CompareExchange(ref hashTableLocal[tableIndex], null, _entryInProcessOfWritingSentinel) == _entryInProcessOfWritingSentinel)
```

Comment:
>Interlocked.CompareExchange [](start = 16, length = 27)

This could be an ordinary write. Unless you specifically do this for the assert in the caller.